### ISSUE TITLE
Add Notes to a Highlight

### ISSUE THREAD

**Issue by hebertialmeida** [FolioReader/FolioReaderKit#16]
Ability to add notes to a Highlight like in iBooks.

**Comment by kjantzer** [comment]
:+1:

**Comment by kjantzer** [comment]
Before we do this, we should probably fix #37

**Comment by hebertialmeida** [comment]
I will design this.

**Comment by hebertialmeida** [comment]
**Concept:** All notes are linked to a highlight, there is no possible to create a note without a highlight, basically if you add a new note it adds a highlight too, or if you have a existing highlight you can add a note to it. This is the same iBooks approach.

**Design:** Simple but I believe that it functional.

### The menu and the _Note_ mark.
![highlght notes](https://cloud.githubusercontent.com/assets/103670/12644959/b3256fee-c5ad-11e5-8f5f-e2e8dd94ae68.png)

### Tap to add or edit a note.
![note edit](https://cloud.githubusercontent.com/assets/103670/12644965/bb327dc6-c5ad-11e5-9e29-e5800f167493.png)


### Tap to open a note, if you tap on content the keyboard opens like on the screen above.
![note preview](https://cloud.githubusercontent.com/assets/103670/12644968/bffa2048-c5ad-11e5-8119-850ee70a84af.png)

### Adding notes we need to change the Highlight list design to also show notes.
![highlight-list](https://cloud.githubusercontent.com/assets/103670/12644972/c9292934-c5ad-11e5-8716-09df7644c747.png)

This is the concept. What do you think @kjantzer

**Comment by kjantzer** [comment]
@hebertialmeida - This looks great!

It's exactly what I was thinking. A while back I looked at how iBooks did notes and came to the same conclusion you did: notes are just a highlight with a user generated note attached. It will make it easier to implement.

The only concern I have is how we'll handle the notes icon next to the highlight when a user adds a note multiple times on the same line :worried: iBooks somehow stacks the notes icons. I was thinking we could use css `:before` to add the note icon to the left side, but now I'm not so sure.

**Comment by hebertialmeida** [comment]
I think You Version App uses `:before` take a look:
![img_4729](https://cloud.githubusercontent.com/assets/103670/12653326/3127f4fc-c5d6-11e5-94b4-652c174657bf.PNG)

Maybe we can do something like this, attached to the highlight.

**Comment by kjantzer** [comment]
Yeah, that might be the easier implementation, at least for the first release.

**Comment by hebertialmeida** [comment]
We can try It and see if works.

On design point of view I prefer on left, but you are right about multiple highlights on same line will be hard to handle.

**Comment by kjantzer** [comment]
I agree, note icon on the left would be preferred. We'll see what is easier to develop

**Comment by tschob** [comment]
Is there any progress on the note implementation? I plan to refactor the Highlights implementation. If there is unmerged code there might be conflicts when https://github.com/FolioReader/FolioReaderKit/issues/131 is implemented.

**Comment by hebertialmeida** [comment]
@tschob No, there is no progress yet, maybe @kjantzer have started working on something. The only thing I have is some designed screens on Sketch. We are stuck on this because we have to refactor the highlights, as it is problematic.

One problem of the current implementation is that you can't highlight text between paragraph tags. We have started a discussion on this on issue #37 and #97 reported this also.

The biggest challenge I think is not just implement this like for a new solution like Rangy which seems to be the best option, mentioned on issue #37, but to migrate the existing highlights from all users.

**Comment by kjantzer** [comment]
@tschob I have not made any significant progress on a highlight refactor so feel free to take a stab at it.

**Comment by benyboariu** [comment]
I need to implement Notes feature, in one of my project. Did anyone start working on this as today?

**Comment by hebertialmeida** [comment]
@benyboariu Not yet. Maybe @tschob did something.

**Comment by tschob** [comment]
@benyboariu Nothing was started by me as well.

**Comment by gabrielvieira** [comment]
@hebertialmeida @tschob  i started to working in this feature by my needs, after implementing the layout of preview i can submit a pull request.

**Comment by onivem** [comment]
Any update with this feature?
